### PR TITLE
Process Tomes

### PR DESCRIPTION
**Pull request by ded8393**
- [x] generate `Item`s from tomes
- [x] generate `Recipe`s from tomes
- [x] process AUTOSAVE data to find "mastered" tomes and add to `known_recipes`
- [x] update AUTOSAVE processing to not include numens in `known_items`

### CODE REVIEW

File: src/boh_app/data/generate_recipes.py [hunk 1]
```diff
@@ -85,3 +84,29 @@ def mk_model_data(self, recipe: dict[str, Any]) -> None:
...
+        for aspect, value in book["aspects"].items():
+            aspect = aspect.removeprefix("mystery.")
+            if aspect in dir(Principle):
+                principle = Principle(aspect)
+                amount = value
+        for name, products in book["xtriggers"].items():
+            if name.startswith("reading."):
+                for p in products:
+                    if p["id"] in self.items:
+                        product = ItemRef(id=p["id"])
```
Comment (flying-sheep) [inline src/boh_app/data/generate_recipes.py:104]: Those are search loops right?

I’d either do

```py
try:
    thing = next(x for xs in y if xs.blah for x in xs)
except StopIteration:
    raise RuntimeError(f"No x found in y: {some_info}")
```

(if not finding things is a valid outcome, just do `thing = next((...), None)`

or, if comprehensions are not powerful enough:

```py
for ...:
    if element.valid:
        thing = element.xy
        break
else:
    raise RuntimeError(...)
```

The `else` branch is hit if one doesn’t `break`

File: src/boh_app/data/process_autosave.py [hunk 2]
```diff
@@ -118,3 +120,18 @@ def get_recipes(self, data: dict[str, Any]) -> list[KnownRecipe]:
...
+        assert root_data["GoverningSphereSpec"]["Id"] == "Library"
+        known_books = set()
+        for lib_token in root_data["Tokens"]:
+            for dominion in lib_token["Payload"]["Dominions"]:
+                for sphere in dominion["Spheres"]:
+                    if sphere["GoverningSphereSpec"]["Label"] == "BOOKSHELF":
+                        for bookshelf_token in sphere["Tokens"]:
+                            payload = bookshelf_token["Payload"]
+                            if any(mut.startswith("mastery.") for mut in payload["Mutations"].keys()):
+                                known_books.add(payload["EntityId"])
```
Comment (flying-sheep) [inline src/boh_app/data/process_autosave.py:135]: you can make this less nested by using a comprehension:

```suggestion
        books = [
            bookshelf_token["Payload"]
            for lib_token in root_data["Tokens"]
            for dominion in lib_token["Payload"]["Dominions"]
            for sphere in dominion["Spheres"]
            if sphere["GoverningSphereSpec"]["Label"] == "BOOKSHELF"
            for bookshelf_token in sphere["Tokens"]
        ]
        known_books = {
            book["EntityId"]
            for book in books
            if any(mut.startswith("mastery.") for mut in payload["Mutations"])
        }
```

**Review by flying-sheep — CHANGES_REQUESTED** [review]
I have a few tricks to get nesting down :D

File: src/boh_app/data/process_autosave.py (hunk 2 above)
Comment (ded8393) [reply src/boh_app/data/process_autosave.py]: mastered_known, unread_known = partition(lambda book: any(mut.startswith("mastery.") for mut in book["Mutations"]), book)

**Review by flying-sheep — APPROVED** [review]

**Pull request merged**